### ISSUE TITLE
X-server crashes on updating xrandr config using muxless nvidia card with external screen

### ISSUE THREAD

**Issue by moritzschaefer** [NixOS/nixpkgs#199024]
### Describe the bug
Using my muxless-GPU Thinkpad with an external screen (no matter whether it's connected to HDMI or USB-C), my XServer crashes when I update my screen-configuration using xrandr/arandr AFTER I enabled a second screen.

### Steps To Reproduce
1. Buy a muxless-GPU laptop (e.g. a modern thinkpad) and install NixOS
2. Plug in a monitor
3. Enable the external port using `xrandr --setprovideroutputsource NVIDIA-G0 modesetting`
4. Enable the external screen (e.g. `xrandr --output DP-1-2.2 --auto`) <- This one still works and the monitor behaves as expected
5. Change the screen configuration (e.g. xrandr --output DP-1-2.2 --off) <- This one crashes the XServer. I have not yet found an xrandr command that works in this situation.

### Expected behavior
- xrandr configuration modifications should work with XServer surviving
- A usable debug output in `/var/log/X.0.log.old` (unfortunately, there is no error/warning message)

### Additional context
Related to previous issues #165188 and #189378

### Notify maintainers
@erplsf, @thiagokokada, @GoogleBot42 

### Metadata

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.72, NixOS, 22.05 (Quokka), 22.05.20221026.e6e675c`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - channels(root): `"nixos-22.05"`
 - channels(moritz): `"nixos-22.05"`
 - nixpkgs: `/nix/store/d5c6h5p16jg2rna7db5y4s3y19dmwvx6-source`
```

**Comment by GoogleBot42** [comment]
It looks like you're trying to use reverse prime. Unfortunately, it doesn't work. I have a PR that I haven't worked on for a while now to try to add support.  The crash you're getting is where I'm stuck. If you'd like to look at the investigation that's happened so far, you might want to take a look. https://github.com/NixOS/nixpkgs/pull/165188

**Comment by moritzschaefer** [comment]
@GoogleBot42  Notably, I don't have reverse_sync enabled in my config..

**Comment by GoogleBot42** [comment]
Yes you are correct.  That xrandr command is turning on reverse prime sync.  the `G0` is specific to reverse prime. https://us.download.nvidia.com/XFree86/Linux-x86_64/515.57/README/primerenderoffload.html

**Comment by moritzschaefer** [comment]
Interestingly, synchronization seems to be enabled for all outputs after I enable reverse prime: 

```
[I] moritz@mopad ~/P/helm-bibtex (master)> xrandr --setprovideroutputsource NVIDIA-G0 modesetting
[I] moritz@mopad ~/P/helm-bibtex (master)> xrandr --prop | ag sync
        PRIME Synchronization: 1
        PRIME Synchronization: 1
        PRIME Synchronization: 1
        PRIME Synchronization: 1
        PRIME Synchronization: 1
        PRIME Synchronization: 1
        PRIME Synchronization: 1
        PRIME Synchronization: 1
        PRIME Synchronization: 1
```

Furthermore, NVIDIA-G0 (corresponding to ID=1) is the only value I can provide to setprovideroutputsource.


```
[I] moritz@mopad ~/P/helm-bibtex (master)> xrandr --listproviders
Providers: number : 2
Provider 0: id: 0x49 cap: 0xf, Source Output, Sink Output, Source Offload, Sink Offload crtcs: 4 outputs: 6 associated providers: 1 name:modesetting
Provider 1: id: 0x24e cap: 0x2, Sink Output crtcs: 4 outputs: 9 associated providers: 1 name:NVIDIA-G0
[I] moritz@mopad ~/P/helm-bibtex (master) [1]> xrandr --setprovideroutputsource 0 modesetting
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  140 (RANDR)
  Minor opcode of failed request:  35 (RRSetProviderOutputSource)
  Value in failed request:  0x49
  Serial number of failed request:  16
  Current serial number in output stream:  17
[I] moritz@mopad ~/P/helm-bibtex (master) [1]> xrandr --setprovideroutputsource 1 modesetting
[I] moritz@mopad ~/P/helm-bibtex (master)> xrandr --setprovideroutputsource 2 modesetting
Could not find provider with index 2
```

**Comment by GoogleBot42** [comment]
Yeah, the xorg config needs to be set in a specific way to configure the NVIDIA card.  In this case it's choosing to be ready for reverse prime.  I've read that which one it chooses by default is dependent on what BIOS wants.  Either way, I'm fairly confident you'll need to enable this option if you want to use prime sync. It configures xorg in that that is necessary. https://search.nixos.org/options?channel=22.05&show=hardware.nvidia.prime.sync.enable&from=0&size=50&sort=relevance&type=packages&query=nvidia+sync

**Comment by MagicRB** [comment]
@moritzschaefer has this been fixed on your end? it is rather annoying to have to restart (segfault) Xorg every time I play with the HDMI.

**Comment by moritzschaefer** [comment]
Not that I know of. Did not test in a while

**Comment by MagicRB** [comment]
:( i captured a syscall trace, but looking at it gave up no useful insight, i may try to compile Xorg with debug symbols and attack GDB

**Comment by MagicRB** [comment]
I posted some updates in https://github.com/NixOS/nixpkgs/pull/165188#issuecomment-1574956220, I'd really appreciate any help I can get. I'm way over my head right now. I also opened a forum post on the nvidia forums https://forums.developer.nvidia.com/t/xorg-segmentation-fault-after-setting-the-mode-of-a-reverse-prime-output-twice-nixos/255470.

**Comment by niscolas** [comment]
I don't know if this can help, but I was having a similar problem.
Whenever I ran any `xrandr` command for the second time (example: ran `xrandr --auto` and then wanted just my external display and ran `xrandr <external-display-options...>`) X crashed.
My solution was to add a xserver config to my `configuration.nix`. (This one worked for me)
```
  services.xserver.config = ''
      Section "ServerLayout"
        Identifier "layout"
        Screen "nvidia" 0 0
      EndSection

      Section "Module"
          Load "modesetting"
          Load "glx"
      EndSection

      Section "Device"
        Identifier "nvidia"
        Driver "nvidia"
        BusID "PCI:1:0:0"
        Option "AllowEmptyInitialConfiguration"
      EndSection

      Section "Device"
        Identifier "intel"
        Driver "modesetting"
        Option "AccelMethod" "sna"
      EndSection

      Section "Screen"
        Identifier     "nvidia"
        Device         "nvidia"
        DefaultDepth    24
        Option         "AllowEmptyInitialConfiguration"
        SubSection     "Display"
          Depth       24
          Modes      "nvidia-auto-select"
        EndSubSection
      EndSection

      Section "Screen"
        Identifier "intel"
        Device "intel"
      EndSection
 '';
```

**Comment by moritzschaefer** [comment]
@MagicRB mentioned in the (https://github.com/NixOS/nixpkgs/pull/165188#issuecomment-1798060101)[closed PR]

> yeah, i didnt get anywhere, though i do believe its a glibc issue and recompiling glibc on our side with the right incantation miggt help

Does this resonate with someone here?

**Comment by MagicRB** [comment]
the reason why i believe so, is that if you attach a debugger, it segfault in glibs while trying to resolve the "Xf86PixmapFree" symbol, or something along those lines. it also segfaults not in nvidia or xorg, but inside glibc. the interesting bit is that if you look at  the backtrace its Xorg -> nvidia -> glibc. I also tried performming a full binary load, ie disabling lazy loading but that didnt helo either.

**Comment by MagicRB** [comment]
This is still happening and it is as annoying as ever. I really do not have the slightest clue how to debug this

**Comment by moritzschaefer** [comment]
@MagicRB is there a way to put a bounty on this one?

**Comment by MagicRB** [comment]
I wouldn't have the resources to create a bounty and tbh the moment zfs for 6.7 lands i'll jump ship to NVK and not bother with steaming pile of garbage anymore except for games.

**Comment by MagicRB** [comment]
but it would be nice to have this resolved i do agree

**Comment by MagicRB** [comment]
> I don't know if this can help, but I was having a similar problem. Whenever I ran any `xrandr` command for the second time (example: ran `xrandr --auto` and then wanted just my external display and ran `xrandr <external-display-options...>`) X crashed. My solution was to add a xserver config to my `configuration.nix`. (This one worked for me)
> 
> ```
>   services.xserver.config = ''
>       Section "ServerLayout"
>         Identifier "layout"
>         Screen "nvidia" 0 0
>       EndSection
> 
>       Section "Module"
>           Load "modesetting"
>           Load "glx"
>       EndSection
> 
>       Section "Device"
>         Identifier "nvidia"
>         Driver "nvidia"
>         BusID "PCI:1:0:0"
>         Option "AllowEmptyInitialConfiguration"
>       EndSection
> 
>       Section "Device"
>         Identifier "intel"
>         Driver "modesetting"
>         Option "AccelMethod" "sna"
>       EndSection
> 
>       Section "Screen"
>         Identifier     "nvidia"
>         Device         "nvidia"
>         DefaultDepth    24
>         Option         "AllowEmptyInitialConfiguration"
>         SubSection     "Display"
>           Depth       24
>           Modes      "nvidia-auto-select"
>         EndSubSection
>       EndSection
> 
>       Section "Screen"
>         Identifier "intel"
>         Device "intel"
>       EndSection
>  '';
> ```

the reason why this helps is because it forces the system into prime output no matter what NixOS says so the intel gpu is serving only as display out. Just tried it out and it does indeed help but running always on the nvidia gpu has its own share of problems (like gamescope not working, thank you nvidia, we love you)

**Comment by ivanovsaleksejs** [comment]
Same problem here. After installing nvidia drivers, calling any xrandr command or using arandr to change configuration, xserver session crashes. And it feels like it crashes every second time.

**Comment by dnr** [comment]
I'm hitting this too: Thinkpad P1 gen 3 with a Intel iGPU plus a Quadro T2000, kernel 6.8.6, nvidia drivers 545.29.02.

I'm using offload mode at the moment, but I'd be happy to use prime or reverse prime if it was more stable (I don't even want to use the dGPU at all, it's just that I have to to use external outputs).

**Comment by MagicRB** [comment]
As i understand if you want to use the external outputs you have two option, offload or prime output. Prime output would mewn that everything is rendered on the dgpu, as to whether its more stable, I dont recall. If you can, id recommend switching to nouveau, theyre planning to support older gpus as much as possible and since you only plan to use it for output it should be fine. If you do need some performance out of it, you can manually reclock it by changing the pci powerstate which affects the whole thing as far as i understand it. If you do decide to go down the nouveau route, i recommend joining the nouveau IRC channel and then trying it out and reporting what you run into.

**Comment by srhb** [comment]
Because this is very old, and because I myself fell into some very wrong assumptions, I just want to point out that there are several setups that _will_ just work with reverse PRIME-setups out of the box, but only if you don't use the proprietary nvidia driver.

For instance, I can drive my Lenovo X1 Extreme Gen 5 with a Nvidia GA107M (Geforce RTX 3050 Ti Mobile) just fine, unplugging and re-plugging outputs as I wish, but ONLY with the stock kernel modules (nouveau, nvidiafb) and modesetting for X. With the proprietary driver and/or all sorts of `hardware.nvidia` config, I only get crashes.

This requires zero lines of nixos config as of nixos-unstable today, the defaults are all correct, so if you don't _need_ the proprietary driver, maybe you're like me and just didn't realize the world is sometimes a better place than expected. Try nuking everything :joy_cat:

**Comment by exzombie** [comment]
Does nouveau automatically and completely disable the dGPU when you are not using external outputs? I, too, only need the dGPU for external outputs, and I would gladly ditch the proprietary driver, but the last time I tried this, the dGPU remained turned on. The proprietary driver will turn it off completely, to the point that even the `sensors` command will no longer be able to read the temperature.

Moreover, what is the power consumption when an external monitor is plugged in? With the proprietary driver, I can work for hours without one, but the moment you plug in a monitor, battery life goes *way* down. If nouveau has larger power consumption than the proprietary driver when driving external outputs, I don't think I can use it.

**Comment by srhb** [comment]
> Does nouveau automatically and completely disable the dGPU when you are not using external outputs

Depends what you mean by "completely"

It goes into d3cold until "probed" or "used" (even probing xrandr or lspci sends it to d0 for a short while) -- I don't know the details of how any of this works.

**Comment by MagicRB** [comment]
d3cold should be as far as it can go normally, but yes the nvidia driver in my experience feels a bit more efficient. Im not sure though. Also warning im on a 3060 laptop gpu with GSP so idk how any of this behaves on pre GSP gpus

**Comment by dnr** [comment]
I did try nouveau at first and it worked, but even normal web browsing felt noticeably sluggish (it's a 4K monitor, I guess it's just a lot of pixels to push around).

I didn't know about the pci powerstate thing, though. As I understand it, nouveau runs the dgpu at a fixed low clock speed, and the proprietary driver can adjust it? But you're saying I can just set it to a fixed higher speed and things will be faster? I'm usually plugged in and don't mind some extra power consumption, so that sounds promising for my situation.

**Comment by MagicRB** [comment]
> I did try nouveau at first and it worked, but even normal web browsing felt noticeably sluggish (it's a 4K monitor, I guess it's just a lot of pixels to push around).
> 
> I didn't know about the pci powerstate thing, though. As I understand it, nouveau runs the dgpu at a fixed low clock speed, and the proprietary driver can adjust it? But you're saying I can just set it to a fixed higher speed and things will be faster? I'm usually plugged in and don't mind some extra power consumption, so that sounds promising for my situation.

Yeah so as far as I understand it, if the firmware uploaded is signed the GPU will kick in self reclocking mechanisms and just do it itself. If its not signed you have to do it manually. You want to change the `pstate` somewhere in `/sys`.

But you'd be much better of joining the #nouveau IRC channel, i think its on libera.

**Comment by ivanovsaleksejs** [comment]
Is there any progress? I can confirm the same problem exists in 24.05. Any chance it will ever be fixed?

**Comment by MagicRB** [comment]
Sadly it's unlikely, I've talked to one of the NVidia developers and we've confirmed it's our problem not theirs. It seems like the nvidia driver mamages to get dybamically linked with a library which it should not be linked with, then proceeds to somehow call into it and segfaults. They've confirmed it should *never* call that lib. I'll look at the conversation later and update with more detail.

**Comment by MagicRB** [comment]
And my recommendation, even though this is a answer no one wants to hear, switch nouveau, if you have a 10 series or older GPU, get a new one... They will not get open driver support and NVidia had or will drop them from the mainline driver. They're readying for dropping the proprietary driver completely which cannot work on 10 series and lower.

**Comment by ivanovsaleksejs** [comment]
My GPU is NVIDIA GeForce RTX 3080 Laptop...

**Comment by MagicRB** [comment]
yeah then nouveau and NVK should work fine for you. Performance is still at 60%ish from what I tested but for normal usage it should be good. Do make sure to run Mesa master and latest Linux. I get random segfault in the DRM driver because I have to be on 6.8 due to ZFS.

**Comment by cafkafk** [comment]
> It seems like the nvidia driver mamages to get dybamically linked with a library which it should not be linked with, then proceeds to somehow call into it and segfaults.

@MagicRB what library is this, also do you have more details (email, issue, prs)?

**Comment by niscolas** [comment]
Has anyone made any progress on this one?

**Comment by esp0xdeadbeef** [comment]
I've got the same problem, anyone willing to share a config?
